pallets/parachain-system/src/lib.rs

```diff
@@ -296,6 +299,34 @@ pub mod pallet {
 				weight += T::DbWeight::get().writes(1);
 			}
 
+			let para_head = ParaHead::<T>::take();
+			weight += T::DbWeight::get().reads_writes(1, 1);
+
+			// Update unincluded segment related storage values.
+			if let Some(para_head) = para_head {
+				let dropped: Vec<BlockTracker> = <UnincludedSegment<T>>::mutate(|chain| {
+					let idx =
+						chain.iter().take_while(|block| block.para_head() != &para_head).count();
+					// Drop the block with an included para head too.
+					let idx = chain.len().min(idx + 1);
```

Comment:
This seems inaccurate.

If we have an unincluded segment of `[A, B, C]` and then `B` is included, we'd get `idx = 1 + 1 = 2` and drain `[A, B]` from the front (correct so far).

But then if `B` is still the best included block at the next call here, we'd get `idx = 0 + 1 = 2` and drain `[C]`.

The fix is to only prune the unincluded segment when the best included parablock appears within it and do nothing otherwise.

Reply:
I tried to account for the case of our fork being abandoned. Correct approach would include looking for the parent too, right?

Reply:
So with that in mind, the correct approach is to clear everything up to and including the last included parablock, and to clear nothing if it is not present.

Reply:
[6549be6](https://github.com/paritytech/cumulus/pull/2438/commits/6549be6216b23d34ff3140d0f57d006c6512dfa0)